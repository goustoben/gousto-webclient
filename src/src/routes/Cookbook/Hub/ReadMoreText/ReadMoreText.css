@value ScreenSize1Max (max-width: 375px);
@value ScreenSize2Min (min-width: 376px);
@value ScreenSize2Max (max-width: 475px);
@value ScreenSize3Min (min-width: 476px);
@value ScreenSize3Max (max-width: 600px);
@value ScreenSize4Min (min-width: 601px);
@value ScreenSize4Max (max-width: 767px);
@value ScreenSize5Min (min-width: 768px);

.screenSize1 {
  @media ScreenSize1Max {
    display: none;
  }
}

.screenSize2 {
  @media ScreenSize2Max {
    display: none;
  }
}

.screenSize3 {
  @media ScreenSize3Max {
    display: none;
  }
}

.screenSize4 {
  @media ScreenSize4Max {
    display: none;
  }
}

.readMore1 {
  @media ScreenSize2Min {
    display: none;
  }
}

.readMore2 {
  @media ScreenSize3Min {
    display: none;
  }
}

.readMore3 {
  @media ScreenSize4Min {
    display: none;
  }
}

.readMore4 {
  @media ScreenSize5Min {
    display: none;
  }
}
