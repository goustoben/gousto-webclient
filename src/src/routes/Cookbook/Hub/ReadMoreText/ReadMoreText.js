import React from 'react'
import PropTypes from 'prop-types'

import css from './ReadMoreText.css'

class ReadMoreText extends React.Component {
  constructor() {
    super()

    this.onReadMoreClick = this.onReadMoreClick.bind(this)

    this.state = { isExpanded: false }
  }

  onReadMoreClick(e) {
    e.preventDefault()
    this.setState({ isExpanded: true })
  }

  renderChunks() {
    const { lines, text } = this.props

    const breakpoints = [
      [40, '', css.screenSize1, css.readMore1],
      [55, css.screenSize1, css.screenSize2, css.readMore2],
      [70, css.screenSize2, css.screenSize3, css.readMore3],
      [95, css.screenSize3, css.screenSize4, css.readMore4],
    ]

    // trim whitespaces
    const fullText = text.replace(/\s/gi, ' ')
    let restTextClass = ''
    let readMoreClass = ''
    let start = 0
    const chunks = []

    for (let i = 0; i < breakpoints.length; i++) {
      const [maxLineLength, chunkClassName, restTextClassName, readMoreClassName] = breakpoints[i]
      const maxLength = maxLineLength * lines
      if (fullText.length > start && fullText.length > maxLength) {
        const shortenText = fullText.substr(start, maxLength - start)
        chunks.push(<span key={i} className={chunkClassName}>{shortenText}</span>)
        start = maxLength
        restTextClass = restTextClassName
        readMoreClass = readMoreClassName
      }
    }

    if (fullText.length > start && restTextClass) {
      chunks.push(<span key="rest-text" className={restTextClass}>{fullText.substr(start)}</span>)
    }

    if (readMoreClass) {
      chunks.push(
        <span key="read-more-link" className={readMoreClass}>
          {'â€¦ '}
          {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
          <a href="#" onClick={this.onReadMoreClick}>Read more</a>
        </span>
      )
    }

    return chunks.length ? chunks : text
  }

  render() {
    const { text } = this.props
    const { isExpanded } = this.state

    return (
      <p>{ isExpanded ? text : this.renderChunks() }</p>
    )
  }
}

ReadMoreText.propTypes = {
  text: PropTypes.string.isRequired,
  lines: PropTypes.number,
}

ReadMoreText.defaultProps = {
  lines: 3,
}

export {
  ReadMoreText
}
