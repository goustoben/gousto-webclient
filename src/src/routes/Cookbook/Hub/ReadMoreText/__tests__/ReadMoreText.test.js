import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { ReadMoreText } from '../ReadMoreText'

const createComponent = (text) => <ReadMoreText lines={3} text={text} />

describe('ReadMoreText', () => {
  let wrapper
  const text = 'Succulent, juicy and always delicious, chicken breasts are a versatile meat that compliment a variety '
    + 'of cuisines. A firm fridge favourite, chicken breasts are easy to cook and work in everything from sizzling '
    + 'stir fries and warming curries to soups and sandwiches. So, whether you want to dice it, fry it or find a new '
    + 'way to try it - we’ve got the best chicken breast recipes to tickle your taste buds. Get the most out of this '
    + 'lean meat and learn new cooking techniques to enhance the flavour. Our chicken breast recipes include '
    + 'asian-inspired stir fries, where the meat is hit first to tenderise it, as well as firm staples like '
    + 'stringy pulled chicken in a juicy poutine sauce. It may be a staple to your diet, but this is chicken like '
    + 'you’ve never tasted before.'

  test('should render paragraph', () => {
    wrapper = shallow(createComponent(text))

    expect(wrapper.find('p')).toHaveLength(1)
  })

  test('should not make breakpoints if text is short', () => {
    const shortText = text.substr(0, 30)

    wrapper = shallow(createComponent(shortText))

    expect(wrapper.html()).toEqual(`<p>${shortText}</p>`)
  })

  it('should create one breakpoint if text does not fit in the first breakpoint boundaries', () => {
    const shortText = text.substr(0, 150)

    const tree = renderer.create(createComponent(shortText)).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should create two breakpoints if text does not fit in the second breakpoint boundaries', () => {
    const shortText = text.substr(0, 200)

    const tree = renderer.create(createComponent(shortText)).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should create three breakpoints if text does not fit in the third breakpoint boundaries', () => {
    const shortText = text.substr(0, 250)

    const tree = renderer.create(createComponent(shortText)).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should create four breakpoints if text does not fit in the forth breakpoint boundaries', () => {
    const tree = renderer.create(createComponent(text)).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should show all text when "read more" is clicked', () => {
    wrapper = shallow(createComponent(text))
    wrapper.find('a').simulate('click', { preventDefault() {} })

    expect(wrapper.html()).toEqual(`<p>${text}</p>`)
  })
})
