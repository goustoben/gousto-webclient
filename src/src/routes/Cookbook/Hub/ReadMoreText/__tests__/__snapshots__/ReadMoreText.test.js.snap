// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadMoreText should create four breakpoints if text does not fit in the forth breakpoint boundaries 1`] = `
<p>
  <span
    className=""
  >
    Succulent, juicy and always delicious, chicken breasts are a versatile meat that compliment a variety of cuisines. A fir
  </span>
  <span
    className="screenSize1"
  >
    m fridge favourite, chicken breasts are easy 
  </span>
  <span
    className="screenSize2"
  >
    to cook and work in everything from sizzling 
  </span>
  <span
    className="screenSize3"
  >
    stir fries and warming curries to soups and sandwiches. So, whether you wan
  </span>
  <span
    className="screenSize4"
  >
    t to dice it, fry it or find a new way to try it - we’ve got the best chicken breast recipes to tickle your taste buds. Get the most out of this lean meat and learn new cooking techniques to enhance the flavour. Our chicken breast recipes include asian-inspired stir fries, where the meat is hit first to tenderise it, as well as firm staples like stringy pulled chicken in a juicy poutine sauce. It may be a staple to your diet, but this is chicken like you’ve never tasted before.
  </span>
  <span
    className="readMore4"
  >
    … 
    <a
      href="#"
      onClick={[Function]}
    >
      Read more
    </a>
  </span>
</p>
`;

exports[`ReadMoreText should create one breakpoint if text does not fit in the first breakpoint boundaries 1`] = `
<p>
  <span
    className=""
  >
    Succulent, juicy and always delicious, chicken breasts are a versatile meat that compliment a variety of cuisines. A fir
  </span>
  <span
    className="screenSize1"
  >
    m fridge favourite, chicken br
  </span>
  <span
    className="readMore1"
  >
    … 
    <a
      href="#"
      onClick={[Function]}
    >
      Read more
    </a>
  </span>
</p>
`;

exports[`ReadMoreText should create three breakpoints if text does not fit in the third breakpoint boundaries 1`] = `
<p>
  <span
    className=""
  >
    Succulent, juicy and always delicious, chicken breasts are a versatile meat that compliment a variety of cuisines. A fir
  </span>
  <span
    className="screenSize1"
  >
    m fridge favourite, chicken breasts are easy 
  </span>
  <span
    className="screenSize2"
  >
    to cook and work in everything from sizzling 
  </span>
  <span
    className="screenSize3"
  >
    stir fries and warming curries to soups 
  </span>
  <span
    className="readMore3"
  >
    … 
    <a
      href="#"
      onClick={[Function]}
    >
      Read more
    </a>
  </span>
</p>
`;

exports[`ReadMoreText should create two breakpoints if text does not fit in the second breakpoint boundaries 1`] = `
<p>
  <span
    className=""
  >
    Succulent, juicy and always delicious, chicken breasts are a versatile meat that compliment a variety of cuisines. A fir
  </span>
  <span
    className="screenSize1"
  >
    m fridge favourite, chicken breasts are easy 
  </span>
  <span
    className="screenSize2"
  >
    to cook and work in everything from
  </span>
  <span
    className="readMore2"
  >
    … 
    <a
      href="#"
      onClick={[Function]}
    >
      Read more
    </a>
  </span>
</p>
`;
